### PR TITLE
feat: test that deploy offchain in the layer 2.

### PR DESCRIPTION
**Pull request by LeonardoVieira1630**
# feat: test that deploys offchain in the layer 2

## All Submissions:

* [X] Have you followed the guidelines in our Contributing document?
* [X] Have you checked to ensure there aren't other open [Pull Requests](../../../pulls) for the same update/change?

## Description
Pull request will add an other test file that deploys the contracts in the Arbitrum local node. This way it will be possibly to test the complete flow of interactions including l2 contracts. 

## Related Issue
#51 
#47 

## Changes
- [X] New feature implementation
- [ ] Bug fix
- [ ] Code refactoring
- [ ] Documentation update
- [ ] Other (please specify)

### Changes to Core Features:

* [X] Have you added an explanation of what your changes do and why you'd like us to include them?
* [X] Have you written new tests for your core changes, as applicable?
* [X] Have you successfully run tests with your changes locally?

### CODE REVIEW

File: packages/client/README [hunk 1]
```diff
@@ -0,0 +1,72 @@
+# Running Tests
+
+This project includes unit tests and end-to-end (E2E) tests. Below are the instructions to run the E2E tests.
+
+## Steps to Run E2E Tests
+
+1. In the terminal, navigate to the contracts directory:
+````cmd
+    cd packages/contracts && forge build
```
Comment (pikonha) [inline packages/client/README:9]: if you add a build script on `contracts/package.json` it could be called like so

```suggestion
    yarn contracts build
```

File: packages/client/README [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+````
+
+2. After building the contracts, return to the root directory:
+````cmd
+    cd .. && cd ..
+````
```
Comment (pikonha) [inline packages/client/README:15]:
```suggestion
```

File: packages/client/README [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+
+3. Start the local Hardhat node:
+````cmd
+    yarn hardhat node
+````
+
+4. With the local Hardhat node running, execute the E2E tests:
+````cmd
+    yarn client hardhat test ./test/l2_e2e.spec.ts --network localhost
```
Comment (pikonha) [inline packages/client/README:24]: this could be a script on the client's package.json as well.

perhaps even following [this example](https://github.com/blockful-io/external-resolver/blob/main/packages/contracts/package.json#L7)

File: packages/client/README [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+````
+
+    Make sure to edit the command as necessary to adjust the test file path and network, if applicable.
+
+
+
+
+
+# Running Tests with Local Arbitrum Node
```
Comment (pikonha) [inline packages/client/README:33]:
```suggestion
## Running Tests with Local Arbitrum Node
```

File: packages/client/README [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+````bash
+./test-node.bash --init
+````
+
+2. After deploying the local node, open another terminal and clone the external resolver from Blockful:
+
+````bash
+git clone https://github.com/blockful-io/external-resolver.git
+`````
+
```
Comment (pikonha) [inline packages/client/README:65]: this commentary seems irrelevant I'd say
```suggestion
```

File: packages/client/README [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+3. Finally, run the tests:
+
+````bash
+yarn client hardhat test ./test/l2_e2e.spec.ts --network localhost
```
Comment (pikonha) [inline packages/client/README:69]: you could turn it into a script as well

File: packages/client/test/l2_e2e.spec.ts [hunk 7]
```diff
@@ -0,0 +1,159 @@
...
+  off-chain information during the tests. It's important to note that this initial test script only sets up the
+  environment and stops at the gateway call. It still requires implementing the connection between the gateway and 
+  layer two, or the gateway and the database.
+*/
+
+// Importing abi and bytecode from contracts folder
+import {
+  abi as abiOffchainResolver,
+  bytecode as bytecodeOffchainResolver,
+} from '../../contracts/out/OffchainResolver.sol/OffchainResolver.json'
```
Comment (pikonha) [inline packages/client/test/l2_e2e.spec.ts:14]:
```suggestion
} from '@blockful/contracts/out/OffchainResolver.sol/OffchainResolver.json'
```

File: packages/client/test/l2_e2e.spec.ts [hunk 8]
```diff
@@ -0,0 +1,159 @@
...
+  bytecode as bytecodeUniversalResolver,
+} from '../../contracts/out/UniversalResolver.sol/UniversalResolver.json'
+
+import { ethers as eth } from 'hardhat'
+import { BaseContract, Contract, Wallet, HashZero } from 'ethers'
+import { normalize, labelhash, namehash } from 'viem/ens'
+import { localhost } from 'viem/chains'
+import { createTestClient, http, publicActions } from 'viem'
+import { HardhatEthersSigner } from '@nomicfoundation/hardhat-ethers/signers'
+// import { expect } from 'chai'
```
Comment (pikonha) [inline packages/client/test/l2_e2e.spec.ts:30]:
```suggestion
```

File: packages/client/test/l2_e2e.spec.ts
Comment (pikonha) [inline packages/client/test/l2_e2e.spec.ts:1]: maybe this file could be called `arbitrum_e2e.spec.ts` since we might want to integrate other L2's in the near future

File: packages/client/test/l2_e2e.spec.ts [hunk 9]
```diff
@@ -0,0 +1,159 @@
...
+
+// Creating an example of Bytes32 variable to represent the Node.
+const root = eth.ZeroHash
+
+let offchainResolver: Contract
+let registry: Contract
+let universalResolver: Contract
+let UniversalResolverContract: BaseContract
+const gatewayUrl = 'http://127.0.0.1:3001'
```
Comment (pikonha) [inline packages/client/test/l2_e2e.spec.ts:39]: the port was changed to `3000`

File: packages/client/test/l2_e2e.spec.ts [hunk 10]
```diff
@@ -0,0 +1,159 @@
...
+    10000000,
+  )
+  await (registry.connect(l1Signer) as Contract).setSubnodeRecord(
+    namehash('eth'),
+    labelhash('offchain'),
+    l1Signer,
+    await offchainResolver.getAddress(),
+    10000000,
+  )
+}
```
Comment (pikonha) [inline packages/client/test/l2_e2e.spec.ts:98]: this could be shared with the other `e2e` spec file

File: packages/client/test/l2_e2e.spec.ts [hunk 11]
```diff
@@ -0,0 +1,159 @@
...
+    //     await client.getEnsAvatar({
+    //       name: ensAddress,
+    //       universalResolverAddress:
+    //         (await UniversalResolverContract.getAddress()) as `0x${string}`,
+    //     })
+    //   } catch (error) {
+    //     const customError = error as BaseError
+    //     console.log(customError.message)
+    //     // expect(customError.message).contain('HTTP request failed')
+    //   }
```
Comment (pikonha) [inline packages/client/test/l2_e2e.spec.ts:157]: is this commented out on purpose?

File: packages/client/test/l2_e2e.spec.ts [hunk 12]
```diff
@@ -0,0 +1,159 @@
...
+  /*
+   This test will fail when it reach the gateway
+   because the offchain part is not implemented yet.
+  */
+  it('Call ENS flow with viem.', async () => {
+    // ENS address
+    const ensAddress = normalize('public.eth')
+    // Getting Avatar from the ens address
+    const client = createTestClient({
+      chain: localhost,
```
Comment (pikonha) [inline packages/client/test/l2_e2e.spec.ts:141]: shouldn't that be `arbitrum`?

**Review by pikonha — CHANGES_REQUESTED** [review]

File: packages/client/README (hunk 6 above)
Comment (LeonardoVieira1630) [reply packages/client/README:69]: I tryed to use a script, but it was not possible to run the node and the tests in parallel. Maybe we can work on this improviment later.

File: packages/client/test/l2_e2e.spec.ts
Comment (LeonardoVieira1630) [reply packages/client/test/l2_e2e.spec.ts:1]: Nice call.

**Comment by pikonha** [comment]
the client's `package.json` is still using `vitest` for running the tests

File: packages/client/test/l2_e2e.spec.ts (hunk 11 above)
Comment (LeonardoVieira1630) [reply packages/client/test/l2_e2e.spec.ts:157]: It will be used after the gatway_l2 implementation. That is why it was not removed.

File: packages/client/test/l2_e2e.spec.ts (hunk 8 above)
Comment (LeonardoVieira1630) [reply packages/client/test/l2_e2e.spec.ts:30]: It will be used after the gatway_l2 implementation. That is why it was not removed.

File: packages/client/test/utils.ts
Comment (pikonha) [inline packages/client/test/utils.ts:1]: rename this file to `deploys.ts`

File: packages/client/test/arbitrum_e2e.spec.ts [hunk 13]
```diff
@@ -0,0 +1,84 @@
...
+  deployOffchainResolver,
+  deployRegistry,
+  deployUniversalResolver,
+} from './utils'
+import { ethers as eth } from 'hardhat'
+import { Wallet } from 'ethers'
+import { normalize } from 'viem/ens'
+import { localhost } from 'viem/chains'
+import { createTestClient, http, publicActions } from 'viem'
+const gatewayUrl = 'http://127.0.0.1:3000'
```
Comment (pikonha) [inline packages/client/test/arbitrum_e2e.spec.ts:21]:
```suggestion
const gatewayUrl = 'http://127.0.0.1:3000/{sender}/{data}.json'
```

File: packages/client/test/arbitrum_e2e.spec.ts [hunk 14]
```diff
@@ -0,0 +1,84 @@
...
+const l1Provider = new eth.JsonRpcProvider('http://127.0.0.1:8545')
+const devAccountPrivateKey =
+  '0xb6b15c8cb491557369f3c7d2c287b053eb229daa9c22138887752191c9520659'
+const devAccountPublicKey = '0x3f1Eae7D46d88F08fc2F8ed27FCb2AB183EB2d0E'
+
+describe('SeuContrato', () => {
+  before(async () => {
+    const signers = await eth.getSigners()
+
+    const walletForL2 = new Wallet(devAccountPrivateKey)
```
Comment (pikonha) [inline packages/client/test/arbitrum_e2e.spec.ts:34]: try using viem instead: https://viem.sh/docs/ethers-migration#viem-11

File: packages/client/test/arbitrum_e2e.spec.ts [hunk 15]
```diff
@@ -0,0 +1,84 @@
...
+} from './utils'
+import { ethers as eth } from 'hardhat'
+import { Wallet } from 'ethers'
+import { normalize } from 'viem/ens'
+import { localhost } from 'viem/chains'
+import { createTestClient, http, publicActions } from 'viem'
+const gatewayUrl = 'http://127.0.0.1:3000'
+
+// Providers
+const l2Provider = new eth.JsonRpcProvider('http://127.0.0.1:8547')
```
Comment (pikonha) [inline packages/client/test/arbitrum_e2e.spec.ts:24]: try using viem instead: https://viem.sh/docs/ethers-migration#viem-1

File: packages/client/test/arbitrum_e2e.spec.ts [hunk 16]
```diff
@@ -0,0 +1,84 @@
...
+      gatewayUrl,
+    )
+  })
+
+  /*
+   This test will fail when it reach the gateway
+   because the offchain part is not implemented yet.
+  */
+  it('Call ENS flow with viem.', async () => {
+    // ENS address
```
Comment (pikonha) [inline packages/client/test/arbitrum_e2e.spec.ts:62]:
```suggestion
```

File: packages/client/test/arbitrum_e2e.spec.ts [hunk 17]
```diff
@@ -0,0 +1,84 @@
...
+  before(async () => {
+    const signers = await eth.getSigners()
+
+    const walletForL2 = new Wallet(devAccountPrivateKey)
+    const l2Signer = walletForL2.connect(l2Provider)
+    const l1Signer = await l1Provider.getSigner(devAccountPublicKey)
+
+    // Deploying the contracts
+    const offchainResolverContract = await deployOffchainResolver(
+      l2Signer,
```
Comment (pikonha) [inline packages/client/test/arbitrum_e2e.spec.ts:40]: I think this should be deployed on L1 and the PublicResolver on the L2

**Review by pikonha — CHANGES_REQUESTED** [review]
